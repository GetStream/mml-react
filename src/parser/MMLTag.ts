import { XmlElement, XmlText } from '@rgrove/parse-xml';

function isXmlElement(node: XmlElement | XmlText): node is XmlElement {
  return !!(node as XmlElement).children;
}

/**
 * MMLTag - Xml tag converted to this MMLTag
 */
export class MMLTag {
  name: string;
  node: XmlElement | XmlText;
  attributes: Record<string, string>;
  children?: MMLTag[];
  key?: string;

  constructor(name: string, node: XmlElement | XmlText, children?: MMLTag[]) {
    this.name = name;
    this.node = node;
    this.attributes = (this.node as XmlElement).attributes || {};
    this.children = children;
  }

  getText() {
    if (this.node.type === 'text') return this.node.text;
    else if (isXmlElement(this.node) && this.node.children.length) return (this.node.children[0] as XmlText).text;
    return '';
  }

  initialState() {
    const { name, value } = this.attributes;
    if (name) return { [name]: value };
    return {};
  }
}
