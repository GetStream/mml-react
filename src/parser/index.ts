export * from './parser';
export * from './MMLTag';
export * from './tree';
export * from './converters';
