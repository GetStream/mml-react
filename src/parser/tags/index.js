import { AddToCalendar, ButtonList, Button, Input, Scheduler, Number } from './data';

import { Carousel, Item, Row, Col } from './layout';

import { Icon, Image, MD, Text } from './core';

export const tags = {
  add_to_calendar: AddToCalendar,
  button_list: ButtonList,
  button: Button,
  input: Input,
  scheduler: Scheduler,
  carousel: Carousel,
  col: Col,
  item: Item,
  row: Row,
  icon: Icon,
  image: Image,
  md: MD,
  text: Text,
  number: Number,
};
