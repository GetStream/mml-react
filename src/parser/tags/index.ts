export * from './base';
export * from './tag';
