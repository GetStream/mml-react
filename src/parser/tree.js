import { converterConfig as ReactConverterConfig } from '../components/converterConfig';

/**
 * Tree - The tree object for MML tags
 */
export class Tree {
  constructor(node, children, converterConfig) {
    this.node = node;
    this.name = node.attributes.name;
    this.children = children;
    this.converterConfig = converterConfig || ReactConverterConfig;
  }

  /**
   * Recursively convert nodes to react
   */
  toReact(parentTag = {}) {
    const reactNodes = [];

    const childTags = parentTag.children || [];
    let count = 0;
    for (const childTag of childTags) {
      count += 1;
      const converter = this.converterConfig[childTag.tagName];
      if (!converter) {
        throw Error(
          `Converter not found for tag ${childTag.tagName}, Available converters are ${Object.keys(
            this.converterConfig,
          )}`,
        );
      }
      childTag.key = `tag-${childTag.tagName}-position-${count}`;
      const children = this.toReact(childTag);
      const reactNode = converter(childTag, children);

      reactNodes.push(reactNode);
    }

    return reactNodes;
  }

  hasData() {
    let data = false;

    function checkForData(nodes) {
      for (const n of nodes) {
        if (n.constructor.data) {
          data = true;
        }
        if (n.children) {
          checkForData(n.children);
        }
      }
    }
    checkForData([this]);
    return data;
  }

  validateTree() {
    const errors = [];

    function runValidation(nodes) {
      for (const n of nodes) {
        const nodeErrors = n.validate();
        errors.push(...nodeErrors);
        if (n.children) {
          runValidation(n.children);
        }
      }
    }
    runValidation([this]);
    return errors;
  }

  validate() {
    const errors = [];
    if (!this.children || this.children.length === 0) {
      errors.append('mml tag is empty');
    }
    return errors;
  }

  initialState() {
    // get initial state for all children
    let state = {};

    if (this.name) {
      state.mml_name = this.name;
    }

    function gatherState(nodes) {
      for (const n of nodes) {
        state = { ...state, ...n.initialState() };
        if (n.children) {
          gatherState(n.children);
        }
      }
    }
    gatherState(this.children);
    return state;
  }
}
