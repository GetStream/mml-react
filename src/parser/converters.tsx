import React, { ReactElement } from 'react';

import { MMLTag } from './MMLTag';
import {
  AddToCalendar,
  Button,
  ButtonList,
  Carousel,
  CarouselItem,
  Col,
  Icon,
  Image,
  Input,
  MD,
  Scheduler,
  Text,
  Row,
  Number,
} from '../components';

export type ConvertorType = (tag: MMLTag, children?: ReactElement[]) => ReactElement;

/**
 * The converters maps MML tags to react nodes
 * Every converter is passed the MML tag as well as the children (as React nodes)
 */
export const converters = {
  button: (tag: MMLTag) => {
    return (
      <Button
        {...tag.attributes}
        key={tag.key}
        text={tag.getText()}
        name={tag.attributes.name}
        value={tag.attributes.value}
      />
    );
  },
  button_list: (tag: MMLTag, children?: JSX.Element[]) => {
    return (
      <ButtonList {...tag.attributes} key={tag.key}>
        {children}
      </ButtonList>
    );
  },
  input: (tag: MMLTag) => {
    return <Input {...tag.attributes} key={tag.key} name={tag.attributes.name} value={tag.attributes.value} />;
  },
  add_to_calendar: (tag: MMLTag) => {
    return (
      <AddToCalendar
        {...tag.attributes}
        key={tag.key}
        title={tag.attributes.title}
        start={tag.attributes.start}
        end={tag.attributes.end}
      />
    );
  },
  col: (tag: MMLTag, children?: JSX.Element[]) => {
    return (
      <Col {...tag.attributes} key={tag.key}>
        {children}
      </Col>
    );
  },
  row: (tag: MMLTag, children?: JSX.Element[]) => {
    return (
      <Row {...tag.attributes} key={tag.key}>
        {children}
      </Row>
    );
  },
  icon: (tag: MMLTag) => {
    return <Icon {...tag.attributes} key={tag.key} name={tag.attributes.name} />;
  },
  image: (tag: MMLTag) => {
    return <Image {...tag.attributes} key={tag.key} src={tag.attributes.src} />;
  },
  md: (tag: MMLTag) => {
    return <MD {...tag.attributes} key={tag.key} text={tag.getText()} />;
  },
  text: (tag: MMLTag) => {
    return <Text {...tag.attributes} key={tag.key} text={tag.getText()} />;
  },
  scheduler: (tag: MMLTag) => {
    return (
      <Scheduler
        {...tag.attributes}
        key={tag.key}
        name={tag.attributes.name}
        dateInterval={parseInt(tag.attributes.dateInterval, 10) || 1} // default to 1 day
        timeInterval={parseInt(tag.attributes.timeInterval, 10) || 30} // default to 30 minutes
        duration={parseInt(tag.attributes.duration, 10) || 30} // default to 30 minutes
        selected={tag.attributes.selected}
        fullDay={tag.attributes.fullDay === 'true'}
        icalAvailability={tag.attributes.icalAvailability}
      />
    );
  },
  carousel: (tag: MMLTag, children?: JSX.Element[]) => {
    return (
      <Carousel slideWidth={tag.attributes.slideWidth} {...tag.attributes} key={tag.key}>
        {children}
      </Carousel>
    );
  },
  item: (tag: MMLTag, children?: JSX.Element[]) => {
    return (
      <CarouselItem {...tag.attributes} key={tag.key}>
        {children}
      </CarouselItem>
    );
  },
  number: (tag: MMLTag) => {
    return <Number {...tag.attributes} key={tag.key} name={tag.attributes.name} value={tag.attributes.value} />;
  },
};
