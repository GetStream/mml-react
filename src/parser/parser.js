import parseXml from '@rgrove/parse-xml';
import { tags } from './tags';
import { Tree } from './tree';

/**
 * SourceToXML - Takes an MML string and converts it to XML nodes
 *
 * @param {string} source MML tag string
 *
 * @returns {array} an Array of XML nodes
 */
export function SourceToXML(source) {
  let src = source;
  // the wrapping MML tags are optional, for parsing simplicity we automatically add them if they are not already there
  if (!src.startsWith('<mml')) src = `<mml>${source}</mml>`;

  // emulate HTML handling of & escaping
  const unescapedAmps = /&(?!amp;|lt;|gt;)/g;
  src = src.replace(unescapedAmps, '&amp;');

  // convert the string to XML nodes
  // this library is relatively lightweight and doesn't do a ton of validation
  return [parseXml(src)];
}

/**
 * XMLtoMMLTree - Takes an array of XML nodes and converts it into an MML Tree
 *
 * @param {type} XMLNodes an array of XML nodes
 *
 * @returns {MMLTree} The MML tree
 */
export function XMLtoMMLTree(XMLNodes) {
  let tree;

  function convertNodes(nodes) {
    const MMLNodes = [];
    for (const n of nodes) {
      let children;
      if (n.children) {
        children = convertNodes(n.children);
      }

      // structured way of looking up mml tags...
      let tagName = n.name;
      if (n.name === 'mml') {
        tree = new Tree(n, children);
        continue;
      }
      // skip the document level element...
      if (n.type === 'document') {
        return children;
      }
      if (n.type === 'text') {
        if (n.text.trim().length > 0) {
          tagName = 'text';
        } else {
          // skip empty text nodes
          continue;
        }
      }

      const TagClass = tags[tagName];

      if (TagClass) {
        const tag = new TagClass(tagName, n, children);
        MMLNodes.push(tag);
      } else {
        console.log('unrecognized element', tagName, Object.keys(tags));
      }
    }
    return MMLNodes;
  }
  convertNodes(XMLNodes);

  return tree;
}

/**
 * Takes an MML string and returns an MML Tree
 *
 * @param {string} source MML tag string
 *
 *  @returns {Tree} An MML Tree
 */
export function Parse(source) {
  const XMLNodes = SourceToXML(source);
  return XMLtoMMLTree(XMLNodes);
}
