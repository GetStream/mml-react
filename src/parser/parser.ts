import parseXml from '@rgrove/parse-xml';

import { ConvertorType } from './converters';
import { MMLTag } from './MMLTag';
import { Tree } from './tree';

/**
 * SourceToXML - Takes an MML string and converts it to XML nodes
 *
 * @param {string} source MML tag string
 *
 * @returns {array} an Array of XML nodes
 */
export function SourceToXML(source: string) {
  let src = source.trim();
  // the wrapping MML tags are optional, for parsing simplicity we automatically add them if they are not already there
  if (!src.startsWith('<mml')) src = `<mml>${source}</mml>`;

  // emulate HTML handling of & escaping
  const unescapedAmps = /&(?!amp;|lt;|gt;)/g;
  src = src.replace(unescapedAmps, '&amp;');

  // convert the string to XML nodes
  // this library is relatively lightweight and doesn't do a ton of validation
  return parseXml(src);
}

function convertNodes(nodes: parseXml.XmlNode[]) {
  return nodes.reduce((acc: MMLTag[], node: parseXml.XmlNode) => {
    const element = node as parseXml.XmlElement;
    let children;
    if (element.children) children = convertNodes(element.children);

    let { name } = element;
    if ((node as parseXml.XmlText).type === 'text') {
      if ((node as parseXml.XmlText).text.trim().length) name = 'text';
      else return acc; // skip empty text elements
    }

    acc.push(new MMLTag(name, node as parseXml.XmlElement | parseXml.XmlText, children));
    return acc;
  }, []);
}

/**
 * XMLtoMMLTree - Takes an array of XML nodes and converts it into an MML Tree
 *
 * @param {type} XMLNodes an array of XML nodes
 *
 * @returns {MMLTree} The MML tree
 */
export function XMLtoMMLTree(document: parseXml.XmlDocument, customConvertors?: Record<string, ConvertorType>) {
  if (!document || !document.children || !document.children.length) throw new Error('bad input');

  const mmlNode = document.children[0] as parseXml.XmlElement;
  if (mmlNode.name !== 'mml') throw new Error('missing mml tag');

  return new Tree(mmlNode, convertNodes(mmlNode.children), customConvertors);
}

/**
 * Takes an MML string and returns an MML Tree
 *
 * @param {string} source MML tag string
 *
 *  @returns {Tree} An MML Tree
 */
export function Parse(source: string, customConvertors?: Record<string, ConvertorType>) {
  const XMLNodes = SourceToXML(source);
  return XMLtoMMLTree(XMLNodes, customConvertors);
}
