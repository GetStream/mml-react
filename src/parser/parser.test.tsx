import React from 'react';
import renderer from 'react-test-renderer';
import { SourceToXML, XMLtoMMLTree } from './parser';

test('mml name and simple text field', () => {
  const mml = '<mml name="john">hi</mml>';
  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.name).toEqual('john');
  expect(tree.type).toBeUndefined();

  expect(tree.children.length).toBe(1);
  expect(tree.reactElements).toMatchSnapshot();

  expect(renderer.create(<>{tree.reactElements}</>).toJSON()).toMatchSnapshot();
});

test('mml with type', () => {
  const mml = '<mml type="card">hi</mml>';
  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.type).toEqual('card');
  expect(tree.children.length).toBe(1);
  expect(tree.reactElements).toMatchSnapshot();

  expect(renderer.create(<>{tree.reactElements}</>).toJSON()).toMatchSnapshot();
});

test('mml with type', () => {
  const mml = '<mml type="card">hi</mml>';
  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.type).toEqual('card');
  expect(tree.children.length).toBe(1);
  expect(tree.toReact()).toMatchSnapshot();

  expect(renderer.create(<>{tree.toReact()}</>).toJSON()).toMatchSnapshot();
});

test('mml with button', () => {
  const mml = `<mml name="support">
                  <text>It looks like your credit card isn't activated yet, activate it now:</text>
                  <button name="action" value="Activate">Activate Card</button>
               </mml>`;

  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.name).toEqual('support');
  expect(tree.children.length).toBe(2);
  expect(tree.reactElements).toMatchSnapshot();

  expect(renderer.create(<>{tree.reactElements}</>).toJSON()).toMatchSnapshot();
});

test('simple carousel', () => {
  const mml = '<carousel><item>a</item><item>b</item></carousel>';

  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.children.length).toBe(1);
  //@ts-ignore
  expect(tree.children[0].children.length).toBe(2);
  expect(tree.reactElements).toMatchSnapshot();

  expect(renderer.create(<>{tree.reactElements}</>).toJSON()).toMatchSnapshot();
});

test('simple input', () => {
  const mml = '<input name="name" value="John" />';

  const nodes = SourceToXML(mml);
  expect(nodes).toMatchSnapshot();

  const tree = XMLtoMMLTree(nodes);
  expect(tree).toMatchSnapshot();
  expect(tree.children.length).toBe(1);
  expect(tree.reactElements).toMatchSnapshot();

  expect(renderer.create(<>{tree.reactElements}</>).toJSON()).toMatchSnapshot();
});

test('invalid MML', () => {
  const mml = '<input name="test" value=1 />';
  expect(() => SourceToXML(mml)).toThrowError(/Attribute value expected/);
});

test('forbid access to object prototype props in mml', () => {
  const mml = '<constructor />';
  const nodes = SourceToXML(mml);
  expect(() => XMLtoMMLTree(nodes)).toThrowError(/Converter not found for tag constructor.+/);
});
