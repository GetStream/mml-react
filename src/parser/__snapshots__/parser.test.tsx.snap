// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mml name and simple text field 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "name": "john",
      },
      "children": Array [
        Object {
          "text": "hi",
          "type": "text",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`mml name and simple text field 2`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {},
      "children": undefined,
      "key": "tag-text-position-0",
      "name": "text",
      "node": Object {
        "text": "hi",
        "type": "text",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": "john",
  "node": Object {
    "attributes": Object {
      "name": "john",
    },
    "children": Array [
      Object {
        "text": "hi",
        "type": "text",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Text
      text="hi"
    />,
  ],
  "type": undefined,
}
`;

exports[`mml name and simple text field 3`] = `
Array [
  <Text
    text="hi"
  />,
]
`;

exports[`mml name and simple text field 4`] = `
<div
  className="mml-text"
>
  hi
</div>
`;

exports[`mml with button 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "name": "support",
      },
      "children": Array [
        Object {
          "text": "
                  ",
          "type": "text",
        },
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "text": "It looks like your credit card isn't activated yet, activate it now:",
              "type": "text",
            },
          ],
          "name": "text",
          "type": "element",
        },
        Object {
          "text": "
                  ",
          "type": "text",
        },
        Object {
          "attributes": Object {
            "name": "action",
            "value": "Activate",
          },
          "children": Array [
            Object {
              "text": "Activate Card",
              "type": "text",
            },
          ],
          "name": "button",
          "type": "element",
        },
        Object {
          "text": "
               ",
          "type": "text",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`mml with button 2`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {},
      "children": Array [
        MMLTag {
          "attributes": Object {},
          "children": undefined,
          "key": "tag-text-position-0",
          "name": "text",
          "node": Object {
            "text": "It looks like your credit card isn't activated yet, activate it now:",
            "type": "text",
          },
        },
      ],
      "key": "tag-text-position-0",
      "name": "text",
      "node": Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "text": "It looks like your credit card isn't activated yet, activate it now:",
            "type": "text",
          },
        ],
        "name": "text",
        "type": "element",
      },
    },
    MMLTag {
      "attributes": Object {
        "name": "action",
        "value": "Activate",
      },
      "children": Array [
        MMLTag {
          "attributes": Object {},
          "children": undefined,
          "key": "tag-text-position-0",
          "name": "text",
          "node": Object {
            "text": "Activate Card",
            "type": "text",
          },
        },
      ],
      "key": "tag-button-position-1",
      "name": "button",
      "node": Object {
        "attributes": Object {
          "name": "action",
          "value": "Activate",
        },
        "children": Array [
          Object {
            "text": "Activate Card",
            "type": "text",
          },
        ],
        "name": "button",
        "type": "element",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": "support",
  "node": Object {
    "attributes": Object {
      "name": "support",
    },
    "children": Array [
      Object {
        "text": "
                  ",
        "type": "text",
      },
      Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "text": "It looks like your credit card isn't activated yet, activate it now:",
            "type": "text",
          },
        ],
        "name": "text",
        "type": "element",
      },
      Object {
        "text": "
                  ",
        "type": "text",
      },
      Object {
        "attributes": Object {
          "name": "action",
          "value": "Activate",
        },
        "children": Array [
          Object {
            "text": "Activate Card",
            "type": "text",
          },
        ],
        "name": "button",
        "type": "element",
      },
      Object {
        "text": "
               ",
        "type": "text",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Text
      text="It looks like your credit card isn't activated yet, activate it now:"
    />,
    <Button
      name="action"
      text="Activate Card"
      value="Activate"
    />,
  ],
  "type": undefined,
}
`;

exports[`mml with button 3`] = `
Array [
  <Text
    text="It looks like your credit card isn't activated yet, activate it now:"
  />,
  <Button
    name="action"
    text="Activate Card"
    value="Activate"
  />,
]
`;

exports[`mml with button 4`] = `
Array [
  <div
    className="mml-text"
  >
    It looks like your credit card isn't activated yet, activate it now:
  </div>,
  <button
    className="mml-btn  mml-btn--text mml-btn--grounded"
    name="action"
    onClick={[Function]}
    type="submit"
    value="Activate"
  >
    Activate Card
  </button>,
]
`;

exports[`mml with type 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "type": "card",
      },
      "children": Array [
        Object {
          "text": "hi",
          "type": "text",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`mml with type 2`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {},
      "children": undefined,
      "key": "tag-text-position-0",
      "name": "text",
      "node": Object {
        "text": "hi",
        "type": "text",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": undefined,
  "node": Object {
    "attributes": Object {
      "type": "card",
    },
    "children": Array [
      Object {
        "text": "hi",
        "type": "text",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Text
      text="hi"
    />,
  ],
  "type": "card",
}
`;

exports[`mml with type 3`] = `
Array [
  <Text
    text="hi"
  />,
]
`;

exports[`mml with type 4`] = `
<div
  className="mml-text"
>
  hi
</div>
`;

exports[`mml with type 5`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "type": "card",
      },
      "children": Array [
        Object {
          "text": "hi",
          "type": "text",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`mml with type 6`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {},
      "children": undefined,
      "key": "tag-text-position-0",
      "name": "text",
      "node": Object {
        "text": "hi",
        "type": "text",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": undefined,
  "node": Object {
    "attributes": Object {
      "type": "card",
    },
    "children": Array [
      Object {
        "text": "hi",
        "type": "text",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Text
      text="hi"
    />,
  ],
  "type": "card",
}
`;

exports[`mml with type 7`] = `
Array [
  <Text
    text="hi"
  />,
]
`;

exports[`mml with type 8`] = `
<div
  className="mml-text"
>
  hi
</div>
`;

exports[`simple carousel 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {},
              "children": Array [
                Object {
                  "text": "a",
                  "type": "text",
                },
              ],
              "name": "item",
              "type": "element",
            },
            Object {
              "attributes": Object {},
              "children": Array [
                Object {
                  "text": "b",
                  "type": "text",
                },
              ],
              "name": "item",
              "type": "element",
            },
          ],
          "name": "carousel",
          "type": "element",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`simple carousel 2`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {},
      "children": Array [
        MMLTag {
          "attributes": Object {},
          "children": Array [
            MMLTag {
              "attributes": Object {},
              "children": undefined,
              "key": "tag-text-position-0",
              "name": "text",
              "node": Object {
                "text": "a",
                "type": "text",
              },
            },
          ],
          "key": "tag-item-position-0",
          "name": "item",
          "node": Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "a",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
        },
        MMLTag {
          "attributes": Object {},
          "children": Array [
            MMLTag {
              "attributes": Object {},
              "children": undefined,
              "key": "tag-text-position-0",
              "name": "text",
              "node": Object {
                "text": "b",
                "type": "text",
              },
            },
          ],
          "key": "tag-item-position-1",
          "name": "item",
          "node": Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "b",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
        },
      ],
      "key": "tag-carousel-position-0",
      "name": "carousel",
      "node": Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "a",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "b",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
        ],
        "name": "carousel",
        "type": "element",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": undefined,
  "node": Object {
    "attributes": Object {},
    "children": Array [
      Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "a",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "text": "b",
                "type": "text",
              },
            ],
            "name": "item",
            "type": "element",
          },
        ],
        "name": "carousel",
        "type": "element",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Carousel>
      <CarouselItem>
        <Text
          text="a"
        />
      </CarouselItem>
      <CarouselItem>
        <Text
          text="b"
        />
      </CarouselItem>
    </Carousel>,
  ],
  "type": undefined,
}
`;

exports[`simple carousel 3`] = `
Array [
  <Carousel>
    <CarouselItem>
      <Text
        text="a"
      />
    </CarouselItem>
    <CarouselItem>
      <Text
        text="b"
      />
    </CarouselItem>
  </Carousel>,
]
`;

exports[`simple carousel 4`] = `
<div
  className="mml-carousel "
>
  <div
    className="mml-carousel__track"
  >
    <div
      className="mml-carousel__slides"
    >
      <div
        className="mml-carousel-item mml-carousel__slide"
        style={
          Object {
            "flex": "0 0 120px",
            "minWidth": "120px",
          }
        }
      >
        <div
          className="mml-text"
        >
          a
        </div>
      </div>
      <div
        className="mml-carousel-item mml-carousel__slide"
        style={
          Object {
            "flex": "0 0 120px",
            "minWidth": "120px",
          }
        }
      >
        <div
          className="mml-text"
        >
          b
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`simple input 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {
            "name": "name",
            "value": "John",
          },
          "children": Array [],
          "name": "input",
          "type": "element",
        },
      ],
      "isRootNode": true,
      "name": "mml",
      "type": "element",
    },
  ],
  "type": "document",
}
`;

exports[`simple input 2`] = `
Tree {
  "children": Array [
    MMLTag {
      "attributes": Object {
        "name": "name",
        "value": "John",
      },
      "children": Array [],
      "key": "tag-input-position-0",
      "name": "input",
      "node": Object {
        "attributes": Object {
          "name": "name",
          "value": "John",
        },
        "children": Array [],
        "name": "input",
        "type": "element",
      },
    },
  ],
  "converters": Object {
    "add_to_calendar": [Function],
    "button": [Function],
    "button_list": [Function],
    "carousel": [Function],
    "col": [Function],
    "icon": [Function],
    "image": [Function],
    "input": [Function],
    "item": [Function],
    "md": [Function],
    "number": [Function],
    "row": [Function],
    "scheduler": [Function],
    "text": [Function],
  },
  "name": undefined,
  "node": Object {
    "attributes": Object {},
    "children": Array [
      Object {
        "attributes": Object {
          "name": "name",
          "value": "John",
        },
        "children": Array [],
        "name": "input",
        "type": "element",
      },
    ],
    "isRootNode": true,
    "name": "mml",
    "type": "element",
  },
  "reactElements": Array [
    <Input
      name="name"
      value="John"
    />,
  ],
  "type": undefined,
}
`;

exports[`simple input 3`] = `
Array [
  <Input
    name="name"
    value="John"
  />,
]
`;

exports[`simple input 4`] = `
<input
  className="mml-input"
  id="mml-name"
  name="name"
  onChange={[Function]}
  placeholder=""
  type="text"
  value="John"
/>
`;
